src/node_api.cc

```diff
@@ -3554,3 +3557,299 @@ napi_status napi_run_script(napi_env env,
   *result = v8impl::JsValueFromV8LocalValue(script_result.ToLocalChecked());
   return GET_RETURN_STATUS(env);
 }
+
+struct napi_threadsafe_function__ {
+  uv_async_t async;
+  uv_idle_t idle;
+  uv_mutex_t mutex;
+  uv_cond_t cond;
+  std::queue<void*> queue;
+  size_t refcount;
+  bool is_closing;
+  v8impl::Reference *ref;
+  napi_env env;
+  void *thread_finalize_data;
+  napi_finalize thread_finalize_callback;
+  void* hint;
+  bool idle_running;
+  size_t max_queue_size;
+  napi_threadsafe_function_call_js call_js_cb;
+};
+
+static void napi_threadsafe_function_idle_cb(uv_idle_t* uv_idle) {
+  void* data;
+  bool popped_value = false;
+  bool delete_function = false;
+  size_t size;
+  napi_value js_cb;
+  napi_threadsafe_function ts_fn =
+      node::ContainerOf(&napi_threadsafe_function__::idle, uv_idle);
+
+  uv_mutex_lock(&ts_fn->mutex);
+  size = ts_fn->queue.size();
+  if (size > 0) {
+    data = ts_fn->queue.front();
+    ts_fn->queue.pop();
+    if (size == ts_fn->max_queue_size && ts_fn->max_queue_size > 0) {
+      uv_cond_signal(&ts_fn->cond);
+    }
+    popped_value = true;
+    size--;
+  }
+  if (size == 0 && ts_fn->refcount == 0) {
+    delete_function = true;
+  }
+  uv_mutex_unlock(&ts_fn->mutex);
+
+  v8::HandleScope handle_scope(ts_fn->env->isolate);
+
+  if (popped_value) {
+    js_cb = v8impl::JsValueFromV8LocalValue(ts_fn->ref->Get());
+
+    node::CallbackScope(ts_fn->env->isolate,
+        v8::Local<v8::Object>::Cast(v8impl::V8LocalValueFromJsValue(js_cb)),
+        {0, 0});
```

Comment:
The documentation shouldn’t state that using `napi_make_callback` is unnecessary if we don’t track actually async context…

Or even better, it seems like we could track async context somewhat well? I’m not sure about it on a the conceptual level, but doesn’t it make sense to provide the context that was active when the function was created?

Reply:
We can accept the parameters necessary for a call to `napi_async_init()` and then make that call from `napi_create_threadsafe_function()`.